client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoResultSet.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.jdbc;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.client.ClientSelectedRole;
+import io.trino.client.QueryData;
+import io.trino.client.QueryStatusInfo;
+import io.trino.client.StatementClient;
+import io.trino.client.StatementStats;
+import org.testng.annotations.Test;
+
+import java.time.ZoneId;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.testng.Assert.assertTrue;
+
+public class TestTrinoResultSet
+{
+    @Test(timeOut = 10000)
+    public void testIteratorCancelWhenQueueNotFull()
+            throws Exception
+    {
+        AtomicReference<Thread> thread = new AtomicReference<>();
+        CountDownLatch interruptedButSwallowedLatch = new CountDownLatch(1);
+        MockAsyncIterator<Iterable<List<Object>>> iterator = new MockAsyncIterator<>(
+                new Iterator<Iterable<List<Object>>>()
+                {
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return true;
+                    }
+
+                    @Override
+                    public Iterable<List<Object>> next()
+                    {
+                        thread.compareAndSet(null, Thread.currentThread());
+                        try {
+                            TimeUnit.MILLISECONDS.sleep(1000);
+                        }
+                        catch (InterruptedException e) {
+                            interruptedButSwallowedLatch.countDown();
+                            Thread.currentThread().interrupt();
```

Comment:
if we're pretending to swallow interrupts (as indicated by interruptedButSwallowedLatch), we shouldn't re-interrupt the current thread, should we?

note: not doing `Thread.currentThread().interrupt()` requires a comment that it's intentional

Reply:
removed